File: packages/dart/test/reporter_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:chki18n/chki18n.dart';
import 'package:test/test.dart';

String samplePath(String name) => 'test/samples/$name';

/// Somewhere outside the project, so a failed run leaves nothing behind in it.
Directory outputDir() => Directory.systemTemp.createTempSync('chki18n-report-');

/// The sample every reporter is measured against: one of nearly every check.
Future<Chki18nResult> analyze() => checkTranslationFiles(
  path: samplePath('locales-all-issues'),
  options: const Chki18nOptions(target: 'en'),
);

String render(Chki18nResult result, Chki18nOptions options, {int? width, String? cwd}) =>
    formatResult(
      result,
      resolveOptions(options.copyWith(target: options.target ?? 'en', color: false)).options,
      Chki18nReportInit(width: width, cwd: cwd),
    );

final RegExp ansi = RegExp(r'\x1b\[');

void main() {
  group('reporter options', () {
    test('defaults to a grouped, coloured report and no file', () {
      final options = resolveOptions().options;

      expect(options.reporter, defaultReporter);
      expect(options.groupBy, defaultGroupBy);
      expect(options.output, isNull);
      expect(options.outputReporter, isNull);
      expect(options.color, isTrue);
    });

    test('falls back and says so when a reporter is not one it knows', () {
      final resolved = resolveOptions(
        const Chki18nOptions(text: Chki18nTextOptions(reporter: 'fancy')),
      );

      expect(resolved.options.reporter, defaultReporter);
      expect(resolved.issues.any((issue) => issue.message.contains('fancy')), isTrue);
    });

    test('reads a reporter name whatever its case', () {
      expect(
        resolveOptions(
          const Chki18nOptions(text: Chki18nTextOptions(reporter: 'JSON')),
        ).options.reporter,
        Chki18nReporter.json,
      );
    });

    test('takes the file reporter from the extension', () {
      expect(reporterOfFileName('report.json'), Chki18nReporter.json);
      expect(reporterOfFileName('report.md'), Chki18nReporter.markdown);
      expect(reporterOfFileName('report.txt'), defaultReporter);
      expect(reporterOfFileName('report'), defaultReporter);
    });

    test('lets an explicit reporter override what the extension implies', () {
      final options =
          resolveOptions(
            const Chki18nOptions(output: 'report.json', reporter: Chki18nReporter.list),
          ).options;

      expect(options.outputReporter, Chki18nReporter.list);
    });

    test('uses the extension when no reporter was named', () {
      final options = resolveOptions(const Chki18nOptions(output: 'report.md')).options;

      expect(options.reporter, defaultReporter);
      expect(options.outputReporter, Chki18nReporter.markdown);
    });

    test('reads a width from a number or the string a flag gives it', () {
      expect(resolveOptions(const Chki18nOptions(width: 80)).options.width, 80);
      expect(
        resolveOptions(const Chki18nOptions(text: Chki18nTextOptions(width: '80'))).options.width,
        80,
      );
    });

    test('measures the terminal instead when the width is not a column count', () {
      for (final value in ['abc', '0', '-10']) {
        final resolved = resolveOptions(Chki18nOptions(text: Chki18nTextOptions(width: value)));

        expect(resolved.options.width, isNull);
        expect(
          resolved.issues.any((issue) => issue.message.contains('not a usable `width`')),
          isTrue,
        );
      }
    });

    test('leaves the width unset when nothing asked for one', () {
      expect(resolveOptions().options.width, isNull);
    });
  });

  group('formatResult', () {
    test('renders one section per locale by default', () async {
      final report = render(await analyze(), const Chki18nOptions());

      expect(report, contains(' ko '));
      expect(report, contains('NO_KEY'));
      expect(report, contains('only-en'));
      expect(report, contains('FAIL'));
      // The locale nothing is wrong with is still worth naming.
      expect(report, contains('Clean: en'));
    });

    test('renders one section per check when asked to', () async {
      final report = render(await analyze(), const Chki18nOptions(groupBy: Chki18nGroupBy.code));

      expect(report, contains(' NO_KEY '));
      expect(report, contains('By locale'));
    });

    test('leaves out the colours when they are turned off', () async {
      expect(ansi.hasMatch(render(await analyze(), const Chki18nOptions())), isFalse);
    });

    test('writes one line per issue as a list', () async {
      final result = await analyze();
      final report = render(result, const Chki18nOptions(reporter: Chki18nReporter.list));
      final lines = report.split('\n').where((line) => line.startsWith('ko')).toList();

      expect(lines.length, result.issues.length);
      expect(lines.first, contains('error'));
      expect(lines.first, contains('NO_KEY'));
    });

    test('writes a table per section as Markdown', () async {
      final report = render(
        await analyze(),
        const Chki18nOptions(reporter: Chki18nReporter.markdown),
      );

      expect(report, startsWith('# Translation check'));
      expect(report, contains('## ko'));
      expect(report, contains('| Level | Check'));
    });

    test('pads a Markdown table to one width, counting a wide character as two', () async {
      final report = render(
        await analyze(),
        const Chki18nOptions(reporter: Chki18nReporter.markdown),
      );
      final rows = report.split('\n').where((line) => line.startsWith('|')).toList();
      final widths = {for (final row in rows) displayWidth(row)};

      // The sample holds a Korean value, so equal widths here can only come from
      // counting columns rather than characters.
      expect(rows.any((row) => RegExp('[가-힣]').hasMatch(row)), isTrue);
      expect(widths.length, 1);
    });

    test('hands back the whole result as JSON, unfiltered', () async {
      final result = await analyze();
      final parsed =
          jsonDecode(
                render(result, const Chki18nOptions(reporter: Chki18nReporter.json, warn: false)),
              )
              as Map<String, Object?>;

      expect((parsed['issues']! as List).length, result.issues.length);
      expect(parsed['summary'], result.summary.toJson());
    });

    test('drops the warnings, and their lines with them, on `no-warn`', () async {
      final result = await analyze();
      final report = render(result, const Chki18nOptions(warn: false));

      expect(report.contains('EMPTY_VALUE'), isFalse);
      expect(report, contains('${result.summary.warn} issues not shown'));
      expect(report, contains('NO_KEY'));
    });

    test('leaves out the heading block and the summary on `no-info`', () async {
      final report = render(await analyze(), const Chki18nOptions(info: false));

      expect(report.contains('Compared 11 keys'), isFalse);
      expect(report, contains('NO_KEY'));
    });

    test('lays the report out to the width it is given', () async {
      final report = render(await analyze(), const Chki18nOptions(), width: 60);
      final rules = report.split('\n').where((line) => line.contains('─')).toList();

      expect(rules, isNotEmpty);

      for (final rule in rules) {
        expect(displayWidth(rule), 60);
      }
    });

    test('wraps a description rather than cutting it short', () async {
      final report = render(await analyze(), const Chki18nOptions(), width: 56);

      expect(report, contains('The key exists in the target language but is'));
      expect(report.contains('...'), isFalse);
    });

    test('says so rather than printing nothing when a run is clean', () async {
      final result = await checkTranslationFiles(
        path: samplePath('multiple-translate-files'),
        options: const Chki18nOptions(target: 'en'),
      );

      expect(render(result, const Chki18nOptions()), contains('PASS'));
    });
  });

  group('the GitHub reporter', () {
    /// One issue, so a property can carry the characters a command breaks on.
    Chki18nResult awkward() => buildResult(
      [
        const Chki18nIssue(
          code: Chki18nCheckCode.noKey,
          level: Chki18nLevel.error,
          locale: 'ko',
          key: 'attr.folder',
          group: '',
          targetValue: 'Folder',
          file: '/repo/lo,cales/ko.json',
          message: 'Missing: here, and there.',
        ),
      ],
      resolveOptions(const Chki18nOptions(target: 'en')).options,
      locales: ['en', 'ko'],
      groups: [''],
      keyCount: 1,
    );

    test('writes one workflow command per issue', () async {
      final result = await analyze();
      final report = render(
        result,
        const Chki18nOptions(reporter: Chki18nReporter.github),
        cwd: Directory.current.path,
      );
      final commands = report.split('\n').where((line) => line.startsWith('::')).toList();

      expect(commands.length, result.issues.length);
      expect(commands.first, startsWith('::error '));
      expect(commands.first, contains('title=chki18n NO_KEY'));
      // The path is relative to the working directory, which is what GitHub
      // resolves an annotation against.
      expect(commands.first, contains('file=test/samples/locales-all-issues/ko.json'));
    });

    test('names each severity the way GitHub does', () async {
      final report = render(
        await analyze(),
        const Chki18nOptions(reporter: Chki18nReporter.github),
      );

      expect(report, contains('::warning '));
      expect(report.contains('::warn '), isFalse);
    });

    test('escapes what would otherwise end a command or a property', () {
      final command =
          render(
            awkward(),
            const Chki18nOptions(reporter: Chki18nReporter.github),
            cwd: '/repo',
          ).split('\n').first;

      expect(command, contains('file=lo%2Ccales/ko.json'));
      // The message keeps its punctuation; only a property value may not.
      expect(command, contains('Missing: here, and there.'));
    });

    test('leaves out what the level options hid', () async {
      final report = render(
        await analyze(),
        const Chki18nOptions(reporter: Chki18nReporter.github, warn: false),
      );

      expect(report.contains('::warning '), isFalse);
      expect(report, contains('::error '));
    });
  });

  group('groupIssues', () {
    test('puts the sections that fail the run first', () async {
      final result = await analyze();
      final sections = groupIssues(result.issues, Chki18nGroupBy.code);

      expect(sections.length, greaterThan(1));
      expect(sections.first.counts.error, greaterThan(0));
      expect(
        sections.fold<int>(0, (total, section) => total + section.issues.length),
        result.issues.length,
      );
    });

    test('collects what the axis cannot name into one section', () {
      final sections = groupIssues([
        const Chki18nIssue(
          code: Chki18nCheckCode.invalidOptions,
          level: Chki18nLevel.warn,
          message: 'bad',
        ),
      ], Chki18nGroupBy.locale);

      expect(sections.first.label, '(general)');
    });
  });

  group('column arithmetic', () {
    test('counts a Korean or Japanese character as two columns', () {
      expect(displayWidth('abc'), 3);
      expect(displayWidth('한국어'), 6);
      expect(displayWidth('日本語'), 6);
    });

    test('pads to a column count rather than a character count', () {
      expect(displayWidth(padTo('한국어', 10)), 10);
      expect(displayWidth(padTo('abc', 10)), 10);
    });

    test('cuts to a column count and marks the cut', () {
      expect(truncate('abcdefghij', 5), 'ab...');
      expect(truncate('abc', 5), 'abc');
      expect(displayWidth(truncate('한국어입니다', 7)), lessThanOrEqualTo(7));
    });
  });

  group('writing the report to a file', () {
    test('creates the directory and writes what the extension asked for', () async {
      final directory = outputDir();

      addTearDown(() => directory.deleteSync(recursive: true));

      final file = '${directory.path}/nested/report.json';
      final result = await checkTranslationFiles(
        path: samplePath('locales-all-issues'),
        options: Chki18nOptions(target: 'en', output: file),
      );
      final written = jsonDecode(File(file).readAsStringSync()) as Map<String, Object?>;

      expect((written['summary']! as Map<String, Object?>)['error'], result.summary.error);
      expect((written['issues']! as List).length, result.issues.length);
    });

    test('never writes escape codes into a file', () async {
      final directory = outputDir();

      addTearDown(() => directory.deleteSync(recursive: true));

      final file = '${directory.path}/report.txt';

      await checkTranslationFiles(
        path: samplePath('locales-all-issues'),
        options: Chki18nOptions(target: 'en', output: file, color: true),
      );

      expect(ansi.hasMatch(File(file).readAsStringSync()), isFalse);
    });

    test('fails the run when the report cannot be written', () async {
      final result = await checkTranslationFiles(
        path: samplePath('locales-no-issue'),
        options: Chki18nOptions(
          target: 'en',
          // A path whose parent is a file cannot be created.
          output: '${samplePath('locales-no-issue')}/en.json/report.txt',
        ),
      );

      expect(result.success, isFalse);
      expect(result.issues.any((issue) => issue.message.contains('could not be written')), isTrue);
    });
  });
}

File: packages/javascript/test/reporter.test.ts
import assert from 'node:assert';
import { mkdtemp, readFile, rm } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { describe, it } from 'node:test';
import {
	buildResult,
	checkTranslationFiles,
	DEFAULT_GROUP_BY,
	DEFAULT_REPORTER,
	displayWidth,
	formatResult,
	GROUP_BY,
	groupIssues,
	padTo,
	REPORTER,
	reporterOfFileName,
	resolveOptions,
	truncate
} from '../dist/index.js';
import type { Chki18nOptions, Chki18nResult } from '../dist/index.js';

const samplePath = (name: string) => join('test', 'samples', name);

/** Somewhere outside the project, so a failed run leaves nothing behind in it. */
const outputDir = () => mkdtemp(join(tmpdir(), 'chki18n-report-'));

/** The sample every reporter is measured against: one of nearly every check. */
const analyze = (): Promise<Chki18nResult> =>
	checkTranslationFiles(samplePath('locales-all-issues'), { target: 'en' });

const render = (
	result: Chki18nResult,
	options: Chki18nOptions,
	init: { width?: number; cwd?: string } = {}
) => formatResult(result, resolveOptions({ target: 'en', color: false, ...options }).options, init);

const ANSI = new RegExp('\\u001b\\[');

describe('reporter options', () => {
	it('defaults to a grouped, coloured report and no file', () => {
		const { options } = resolveOptions();

		assert.strictEqual(options.reporter, DEFAULT_REPORTER);
		assert.strictEqual(options.groupBy, DEFAULT_GROUP_BY);
		assert.strictEqual(options.output, null);
		assert.strictEqual(options.outputReporter, null);
		assert.strictEqual(options.color, true);
	});

	it('falls back and says so when a reporter is not one it knows', () => {
		const { options, issues } = resolveOptions({ reporter: 'fancy' as never });

		assert.strictEqual(options.reporter, DEFAULT_REPORTER);
		assert.ok(issues.some((issue) => issue.message.includes('fancy')));
	});

	it('reads a reporter name whatever its case', () => {
		assert.strictEqual(
			resolveOptions({ reporter: 'JSON' as never }).options.reporter,
			REPORTER.JSON
		);
	});

	it('takes the file reporter from the extension', () => {
		assert.strictEqual(reporterOfFileName('report.json'), REPORTER.JSON);
		assert.strictEqual(reporterOfFileName('report.md'), REPORTER.MARKDOWN);
		assert.strictEqual(reporterOfFileName('report.txt'), DEFAULT_REPORTER);
		assert.strictEqual(reporterOfFileName('report'), DEFAULT_REPORTER);
	});

	it('lets an explicit reporter override what the extension implies', () => {
		const { options } = resolveOptions({ output: 'report.json', reporter: REPORTER.LIST });

		assert.strictEqual(options.outputReporter, REPORTER.LIST);
	});

	it('uses the extension when no reporter was named', () => {
		const { options } = resolveOptions({ output: 'report.md' });

		assert.strictEqual(options.reporter, DEFAULT_REPORTER);
		assert.strictEqual(options.outputReporter, REPORTER.MARKDOWN);
	});

	it('reads a width from a number or the string a flag gives it', () => {
		assert.strictEqual(resolveOptions({ width: 80 }).options.width, 80);
		assert.strictEqual(resolveOptions({ width: '80' }).options.width, 80);
	});

	it('measures the terminal instead when the width is not a column count', () => {
		for (const value of ['abc', 0, -10]) {
			const { options, issues } = resolveOptions({ width: value });

			assert.strictEqual(options.width, null);
			assert.ok(issues.some((issue) => issue.message.includes('not a usable `width`')));
		}
	});

	it('leaves the width unset when nothing asked for one', () => {
		assert.strictEqual(resolveOptions().options.width, null);
	});
});

describe('formatResult', () => {
	it('renders one section per locale by default', async () => {
		const report = render(await analyze(), {});

		assert.ok(report.includes(' ko '));
		assert.ok(report.includes('NO_KEY'));
		assert.ok(report.includes('only-en'));
		assert.ok(report.includes('FAIL'));
		// The locale nothing is wrong with is still worth naming.
		assert.ok(report.includes('Clean: en'));
	});

	it('renders one section per check when asked to', async () => {
		const report = render(await analyze(), { groupBy: GROUP_BY.CODE });

		assert.ok(report.includes(' NO_KEY '));
		assert.ok(report.includes('By locale'));
	});

	it('leaves out the colours when they are turned off', async () => {
		assert.strictEqual(ANSI.test(render(await analyze(), {})), false);
	});

	it('writes one line per issue as a list', async () => {
		const result = await analyze();
		const report = render(result, { reporter: REPORTER.LIST });
		const lines = report.split('\n').filter((line) => line.startsWith('ko'));

		assert.strictEqual(lines.length, result.issues.length);
		assert.ok(lines[0].includes('error'));
		assert.ok(lines[0].includes('NO_KEY'));
	});

	it('writes a table per section as Markdown', async () => {
		const report = render(await analyze(), { reporter: REPORTER.MARKDOWN });

		assert.ok(report.startsWith('# Translation check'));
		assert.ok(report.includes('## ko'));
		assert.ok(report.includes('| Level | Check'));
	});

	it('pads a Markdown table to one width, counting a wide character as two', async () => {
		const report = render(await analyze(), { reporter: REPORTER.MARKDOWN });
		const rows = report.split('\n').filter((line) => line.startsWith('|'));
		const widths = new Set(rows.map((row) => displayWidth(row)));

		// The sample holds a Korean value, so equal widths here can only come from
		// counting columns rather than characters.
		assert.ok(rows.some((row) => /[가-힣]/.test(row)));
		assert.strictEqual(widths.size, 1);
	});

	it('hands back the whole result as JSON, unfiltered', async () => {
		const result = await analyze();
		const parsed = JSON.parse(render(result, { reporter: REPORTER.JSON, warn: false }));

		assert.strictEqual(parsed.issues.length, result.issues.length);
		assert.deepStrictEqual(parsed.summary, result.summary);
	});

	it('drops the warnings, and their lines with them, on `no-warn`', async () => {
		const result = await analyze();
		const report = render(result, { warn: false });

		assert.strictEqual(report.includes('EMPTY_VALUE'), false);
		assert.ok(report.includes(`${result.summary.warn} issues not shown`));
		assert.ok(report.includes('NO_KEY'));
	});

	it('leaves out the heading block and the summary on `no-info`', async () => {
		const report = render(await analyze(), { info: false });

		assert.strictEqual(report.includes('Compared 11 keys'), false);
		assert.ok(report.includes('NO_KEY'));
	});

	it('lays the report out to the width it is given', async () => {
		const report = render(await analyze(), {}, { width: 60 });
		const rules = report.split('\n').filter((line) => line.includes('─'));

		assert.ok(rules.length > 0);

		for (const rule of rules) {
			assert.strictEqual(displayWidth(rule), 60);
		}
	});

	it('wraps a description rather than cutting it short', async () => {
		const report = render(await analyze(), {}, { width: 56 });

		assert.ok(report.includes('The key exists in the target language but is'));
		assert.strictEqual(report.includes('...'), false);
	});

	it('says so rather than printing nothing when a run is clean', async () => {
		const result = await checkTranslationFiles(samplePath('multiple-translate-files'), {
			target: 'en'
		});

		assert.ok(render(result, {}).includes('PASS'));
	});
});

describe('the GitHub reporter', () => {
	/** One issue, so a property can carry the characters a command breaks on. */
	const awkward = () =>
		buildResult(
			[
				{
					code: 'NO_KEY',
					level: 'error',
					locale: 'ko',
					key: 'attr.folder',
					group: '',
					targetValue: 'Folder',
					file: '/repo/lo,cales/ko.json',
					message: 'Missing: here, and there.'
				}
			],
			resolveOptions({ target: 'en' }).options,
			{ locales: ['en', 'ko'], groups: [''], keyCount: 1 }
		);

	it('writes one workflow command per issue', async () => {
		const result = await analyze();
		const report = render(result, { reporter: REPORTER.GITHUB }, { cwd: process.cwd() });
		const commands = report.split('\n').filter((line) => line.startsWith('::'));

		assert.strictEqual(commands.length, result.issues.length);
		assert.ok(commands[0].startsWith('::error '));
		assert.ok(commands[0].includes('title=chki18n NO_KEY'));
		// The path is relative to the working directory, which is what GitHub
		// resolves an annotation against.
		assert.ok(commands[0].includes('file=test/samples/locales-all-issues/ko.json'));
	});

	it('names each severity the way GitHub does', async () => {
		const report = render(await analyze(), { reporter: REPORTER.GITHUB });

		assert.ok(report.includes('::warning '));
		assert.strictEqual(report.includes('::warn '), false);
	});

	it('escapes what would otherwise end a command or a property', () => {
		const [command] = render(awkward(), { reporter: REPORTER.GITHUB }, { cwd: '/repo' }).split(
			'\n'
		);

		assert.ok(command.includes('file=lo%2Ccales/ko.json'));
		// The message keeps its punctuation; only a property value may not.
		assert.ok(command.includes('Missing: here, and there.'));
	});

	it('writes a path GitHub can match, whatever the runner calls a separator', () => {
		const onWindows = buildResult(
			[
				{
					code: 'NO_KEY',
					level: 'error',
					locale: 'ko',
					key: 'attr.folder',
					group: '',
					file: 'D:\\repo\\locales\\ko.json',
					message: 'Missing.'
				}
			],
			resolveOptions({ target: 'en' }).options,
			{ locales: ['en', 'ko'], groups: [''], keyCount: 1 }
		);
		const [command] = render(onWindows, { reporter: REPORTER.GITHUB }, { cwd: 'D:\\repo' }).split(
			'\n'
		);

		// GitHub matches an annotation against a repository path, which is always
		// written with forward slashes.
		assert.ok(command.includes('file=locales/ko.json'));
	});

	it('leaves out what the level options hid', async () => {
		const report = render(await analyze(), { reporter: REPORTER.GITHUB, warn: false });

		assert.strictEqual(report.includes('::warning '), false);
		assert.ok(report.includes('::error '));
	});
});

describe('groupIssues', () => {
	it('puts the sections that fail the run first', async () => {
		const result = await analyze();
		const sections = groupIssues(result.issues, GROUP_BY.CODE);

		assert.ok(sections.length > 1);
		assert.ok(sections[0].counts.error > 0);
		assert.strictEqual(
			sections.reduce((total, section) => total + section.issues.length, 0),
			result.issues.length
		);
	});

	it('collects what the axis cannot name into one section', () => {
		const sections = groupIssues(
			[{ code: 'INVALID_OPTIONS', level: 'warn', locale: '', key: '', group: '', message: 'bad' }],
			GROUP_BY.LOCALE
		);

		assert.strictEqual(sections[0].label, '(general)');
	});
});

describe('column arithmetic', () => {
	it('counts a Korean or Japanese character as two columns', () => {
		assert.strictEqual(displayWidth('abc'), 3);
		assert.strictEqual(displayWidth('한국어'), 6);
		assert.strictEqual(displayWidth('日本語'), 6);
	});

	it('pads to a column count rather than a character count', () => {
		assert.strictEqual(displayWidth(padTo('한국어', 10)), 10);
		assert.strictEqual(displayWidth(padTo('abc', 10)), 10);
	});

	it('cuts to a column count and marks the cut', () => {
		assert.strictEqual(truncate('abcdefghij', 5), 'ab...');
		assert.strictEqual(truncate('abc', 5), 'abc');
		assert.ok(displayWidth(truncate('한국어입니다', 7)) <= 7);
	});
});

describe('writing the report to a file', () => {
	it('creates the directory and writes what the extension asked for', async (t) => {
		const directory = await outputDir();

		t.after(() => rm(directory, { force: true, recursive: true }));

		const file = join(directory, 'nested', 'report.json');
		const result = await checkTranslationFiles(samplePath('locales-all-issues'), {
			target: 'en',
			output: file
		});
		const written = JSON.parse(await readFile(file, { encoding: 'utf-8' }));

		assert.strictEqual(written.summary.error, result.summary.error);
		assert.strictEqual(written.issues.length, result.issues.length);
	});

	it('never writes escape codes into a file', async (t) => {
		const directory = await outputDir();

		t.after(() => rm(directory, { force: true, recursive: true }));

		const file = join(directory, 'report.txt');

		await checkTranslationFiles(samplePath('locales-all-issues'), {
			target: 'en',
			output: file,
			color: true
		});

		assert.strictEqual(ANSI.test(await readFile(file, { encoding: 'utf-8' })), false);
	});

	it('fails the run when the report cannot be written', async () => {
		const result = await checkTranslationFiles(samplePath('locales-no-issue'), {
			target: 'en',
			// A path whose parent is a file cannot be created.
			output: join(samplePath('locales-no-issue'), 'en.json', 'report.txt')
		});

		assert.strictEqual(result.success, false);
		assert.ok(result.issues.some((issue) => issue.message.includes('could not be written')));
	});
});
